File: maxima/mergesort.m
mergesort(list) :=
  block([len: length(list), h: entier(length(list)/2)],
    if len > 1 then
      merge(mergesort(rest(list, h-len)), mergesort(rest(list, h)))
    else
      list);

merge(a, b) :=
  if length(a) * length(b) = 0 then
    append(a, b)
  else
    if first(a) < first(b) then
      cons(first(a), merge(rest(a), b))
    else
      cons(first(b), merge(a, rest(b)));

File: ruby/mergesort.rb
def merge(a, b)
  r = []
  r << (a.first < b.first ? a : b).shift until a.empty? || b.empty?
  r + a + b
end

def mergesort(lst)
  return lst if lst.size == 1

  a, b = lst.each_slice((lst.size + 1) / 2).to_a

  merge(mergesort(a), mergesort(b))
end

puts mergesort([10, 9, 8, 5, 6, 7, 4, 3, 2, 1])
